File: src/main/java/AoC2022_08.java
import static java.util.stream.Collectors.toList;

import java.util.List;
import java.util.Set;
import java.util.stream.Stream;

import com.github.pareronia.aoc.IntGrid;
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

public class AoC2022_08 extends AoCBase {
    
    private final IntGrid grid;
    
    private AoC2022_08(final List<String> input, final boolean debug) {
        super(debug);
        this.grid = IntGrid.from(input);
    }
    
    public static final AoC2022_08 create(final List<String> input) {
        return new AoC2022_08(input, false);
    }

    public static final AoC2022_08 createDebug(final List<String> input) {
        return new AoC2022_08(input, true);
    }
    
    public Set<Stream<Cell>> capitalDirections(final Cell cell) {
        return Set.of(
            grid.getCellsN(cell), grid.getCellsE(cell),
            grid.getCellsS(cell), grid.getCellsW(cell));
    }
    
    public Stream<Cell> ignoringBorders() {
        return grid.getCells()
                .filter(cell -> 1 <= cell.getRow())
                .filter(cell -> cell.getRow() < grid.getMaxRowIndex())
                .filter(cell -> 1 <= cell.getCol())
                .filter(cell -> cell.getCol() < grid.getMaxColIndex());
    }
    
    private boolean visibleFromOutside(final Cell cell) {
        final int val = grid.getValue(cell);
        return capitalDirections(cell).stream()
                .anyMatch(dir -> dir.allMatch(c -> grid.getValue(c) < val));
    }
    
    private Integer viewingDistance(final Stream<Cell> direction, final int val) {
        int n = 0;
        boolean stop = false;
        for (final Cell c : direction.collect(toList())) {
            if (stop) {
                break;
            }
            n++;
            stop = grid.getValue(c) >= val;
        }
        return n;
    }
    
    private Integer scenicScore(final Cell cell) {
        return capitalDirections(cell).stream()
                .map(direction -> viewingDistance(direction, grid.getValue(cell)))
                .reduce(1, (a, b) -> a * b);
    }
    
    @Override
    public Integer solvePart1() {
        final int ans = 2 * (grid.getHeight() + grid.getWidth()) - 4;
        return (int) (ans + ignoringBorders().filter(this::visibleFromOutside).count());
    }

    @Override
    public Integer solvePart2() {
        return ignoringBorders().mapToInt(this::scenicScore).max().getAsInt();
    }

    public static void main(final String[] args) throws Exception {
        assert AoC2022_08.createDebug(TEST).solvePart1() == 21;
        assert AoC2022_08.createDebug(TEST).solvePart2() == 8;

        final Puzzle puzzle = Aocd.puzzle(2022, 8);
        final List<String> inputData = puzzle.getInputData();
        puzzle.check(
            () -> lap("Part 1", AoC2022_08.create(inputData)::solvePart1),
            () -> lap("Part 2", AoC2022_08.create(inputData)::solvePart2)
        );
    }

    private static final List<String> TEST = splitLines(
        "30373\r\n" +
        "25512\r\n" +
        "65332\r\n" +
        "33549\r\n" +
        "35390"
    );
}

File: src/main/python/AoC2022_08.py
#! /usr/bin/env python3
#
# Advent of Code 2022 Day 8
#

import sys
from collections.abc import Iterator
from dataclasses import dataclass
from math import prod
from typing import Self

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.grid import Cell
from aoc.grid import IntGrid

TEST = """\
30373
25512
65332
33549
33548
"""


@dataclass(frozen=True)
class Trees:
    grid: IntGrid

    @classmethod
    def from_input(cls, inputs: InputData) -> Self:
        return cls(IntGrid([[int(c) for c in line] for line in inputs]))

    def capital_directions(self, cell: Cell) -> Iterator[Iterator[Cell]]:
        yield self.grid.get_cells_n(cell)
        yield self.grid.get_cells_e(cell)
        yield self.grid.get_cells_s(cell)
        yield self.grid.get_cells_w(cell)

    def count_visible_from_outside(self) -> int:
        return (
            2 * (self.grid.get_height() + self.grid.get_width())
            - 4
            + sum(
                any(
                    all(
                        self.grid.get_value(c) < self.grid.get_value(cell)
                        for c in cells
                    )
                    for cells in self.capital_directions(cell)
                )
                for cell in self.grid.get_cells_without_border()
            )
        )

    def max_scenic_score(self) -> int:
        def viewing_distance(direction: Iterator[Cell], val: int) -> int:
            n = 0
            stop = False
            for c in direction:
                if stop:
                    break
                n += 1
                stop = self.grid.get_value(c) >= val
            return n

        return max(
            prod(
                viewing_distance(direction, self.grid.get_value(cell))
                for direction in self.capital_directions(cell)
            )
            for cell in self.grid.get_cells_without_border()
        )


Input = Trees
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return Trees.from_input(input_data)

    def part_1(self, trees: Input) -> Output1:
        return trees.count_visible_from_outside()

    def part_2(self, trees: Input) -> Output2:
        return trees.max_scenic_score()

    @aoc_samples(
        (
            ("part_1", TEST, 21),
            ("part_2", TEST, 8),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2022, 8)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
